File: frontend/src/workbench/commands.ts
import type { LayoutGetRequest } from "../generated/LayoutGetRequest";
import type { LayoutGetResponse } from "../generated/LayoutGetResponse";
import type { LayoutSetRequest } from "../generated/LayoutSetRequest";
import type { LayoutSetResponse } from "../generated/LayoutSetResponse";
import type { WorkspaceListRequest } from "../generated/WorkspaceListRequest";
import type { WorkspaceListResponse } from "../generated/WorkspaceListResponse";
import type { InvokeOptions, IpcClient } from "../ipc";

export const LAYOUT_COMMANDS = {
  get: "state.layout.get",
  set: "state.layout.set",
  listWorkspaces: "workspace.list",
} as const;

export function listWorkspaces(client: IpcClient, options?: InvokeOptions) {
  return client.invoke<WorkspaceListRequest, WorkspaceListResponse>(
    LAYOUT_COMMANDS.listWorkspaces,
    {},
    options,
  );
}

export function getLayout(client: IpcClient, workspaceKey: string, options?: InvokeOptions) {
  const request: LayoutGetRequest = { workspace_key: workspaceKey };
  return client.invoke<LayoutGetRequest, LayoutGetResponse>(LAYOUT_COMMANDS.get, request, options);
}

export function setLayout(
  client: IpcClient,
  workspaceKey: string,
  layout: unknown,
  options?: InvokeOptions,
) {
  const request: LayoutSetRequest = { workspace_key: workspaceKey, layout };
  return client.invoke<LayoutSetRequest, LayoutSetResponse>(LAYOUT_COMMANDS.set, request, options);
}

File: crates/helix-state/src/commands.rs
//! Typed IPC contract for durable workbench layout state (Task 2.1).

use serde::{Deserialize, Serialize};
use serde_json::Value;
use ts_rs::TS;

pub const LAYOUT_GET: &str = "state.layout.get";
pub const LAYOUT_SET: &str = "state.layout.set";

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
#[serde(default)]
pub struct LayoutGetRequest {
    pub workspace_key: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct LayoutGetResponse {
    #[ts(type = "unknown")]
    pub layout: Value,
    /// Stable content hash used by frontend projection reconciliation.
    pub projection_hash: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
#[serde(default)]
pub struct LayoutSetRequest {
    pub workspace_key: String,
    #[ts(type = "unknown")]
    pub layout: Value,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct LayoutSetResponse {
    pub persisted: bool,
    pub projection_hash: String,
}

pub fn layout_projection_hash(layout: &Value) -> Result<String, serde_json::Error> {
    let bytes = serde_json::to_vec(layout)?;
    Ok(format!("{:08x}", crc32fast::hash(&bytes)))
}
